### ISSUE TITLE
Let targets set what platforms they build for

### ISSUE THREAD

**Issue by gregestren** [bazelbuild/bazel#6519]
Tracking issue on [Bazel Configurability Roadmap](https://bazel.build/roadmaps/configuration.html)

Goal:
* Something like this works:

```
$ cat a/BUILD
cc_binary(name = "app_for_mac", platforms = ["//platforms:mac"])

$ bazel build //a:app_for_mac # Automatically builds for a Mac, no command line flags!
```

This syntax is theoretical; it may not look exactly like this. The point is rules can embed which platforms they target and all the correct toolchains get linked in without having to specify anything at the command line.

Furthermore, *different* rules needing *different* toolchains (e.g. `$ bazel build //a:all`) will all independently do the right thing.

This is v0 because it may have [performance consequences](https://bazel.build/roadmaps/configuration.html#correctness-and-speed). This also doesn't cover non-platform settings (like app version, enable/disable feature X), which will be covered elsewhere in the [roadmap](https://bazel.build/roadmaps/configuration.html).

**Comment by gregestren** [comment]
Prerequisite work is still happening, so marking as a P3 now. This means we're not actively working on this yet. Will upgrade priority once the prereqs are out of the way.

**Comment by gregestren** [comment]
**April '19 update:**

Still blocked on prerequisites of basic platform & toolchain support. I'd still estimate a "second half of the year" time frame for starting to experiment with this, unless there's enough enthusiastic interest to jump on board even without concern about the scaling & caching concerns.

Technically I think really interested users will be able to explore this in the next month or two by manually combining together platform/toolchain support and Starlark transitions.

**Comment by wpieterse2825** [comment]
@gregestren Hi Greg, can you elaborate on how we can achieve multi-platform builds via transitions?

I'm building a small kernel using Bazel as the build engine. The boot-loader is being written targeting UEFI, which uses a PE COFF toolchain (basically targeting Windows via CLang/LLVM), and I want the kernel executables itself to be in ELF format (x86_64-none-elf).

At the moment, I have no clear path to go about it, was thinking more in the lines of having two separate Bazel build steps, and then combining the data into a boot disk, but if I could do everything inside Bazel it will be so much easier.

**Comment by gregestren** [comment]
Hey @wpieterse2825,

Building in separate steps is indeed an option if it comes to that. A major inspiration of these new APIs is to not have to resort to that, so we're on the same page there. 

What BUILD rules are you using to represent the boot-loader and kernel executables and what flags are you using to switch the toolchain?

I suspect we can get a clear path for your use case.

**Comment by wpieterse2825** [comment]
@gregestren Thanks for the reply Greg.

At the moment they are just plain `cc_binary` and `cc_library` rules. You can have a look at the [repository].

I'm not doing anything fancy at the moment, just settings the `--crosstool-top` switch to point to my custom toolchain that can build UEFI applications. I still need to separate out the various pieces, I'm just not sure what is the correct direction forward. What would be you're opinion on how I should handle this? I saw the configuration and build road-map on the Bazel website, and I'm quite excited for the option to use platforms inside the `cc_binary` and `cc_library` rules.

Also, this is just a pet project to build a small "DOS" like operating system, nothing serious.

[repository]: https://dev.azure.com/wynand-pieterse-personal/Paxbox/_git/Master?path=%2Fpaxos&version=GBmaster

**Comment by wpieterse2825** [comment]
Also, to add to the above, the `uefi` part needs to be built with the PE COFF toolchain, while the `kernel` and `architecture` parts I need in ELF format.

Thinking of it, I’m using LLVM and LLD, maybe I should try playing with creating a macro wrapping the cc rules for the UEFI and Kernel parts, where the macro itselfs sets copts to specify what linker to use.

**Comment by katre** [comment]
@wpieterse2825: Have you taken a look at https://github.com/bazelbuild/bazel/issues/7260? Once you write `toolchain()` targets for your existing `cc_toolchain()` declarations, you can use the flag to enable platform-based cc toolchain resolution now, which might be simpler than using `copts` and `crosstool_top`. All of the missing features of full multiplatform builds are still present, of course, but it moves you closer to that world.

**Comment by wpieterse2825** [comment]
Hi @katre, thanks for mentioning it.

I did try and do that, defining the various architectures and firmware targets. You can have a look [here]. But when I try and build only with platforms, without specifying the `--cpu` parameter, it always tries to build with the MSVC toolchain, which obviously causes problems when linking the binary. I could not figure it out at the time, will give it a try again and let you know my results.

[here]: https://dev.azure.com/wynand-pieterse-personal/Paxbox/_git/Master?path=%2Ftools%2Ftoolchain%2Fllvm%2Fuefi%2FBUILD&version=GBmaster

**Comment by wpieterse2825** [comment]
@katra I did what you suggested and it worked. Could finally remove the `--cpu` flag from my `.bazelrc` file as everything is now compiled with platforms. Thanks.

**Comment by katre** [comment]
Excellent, I'm glad it's working, we're trying hard to get this ready enough to enable by default.

**Comment by gregestren** [comment]
As for the multiplatform part, @juliexxia is working on much more extensive documentation. But in the meantime check out https://github.com/gregestren/snippets/wiki/Configuring-a-cc_binary-differently-than-a-cc_test for an example of configuring `cc_` rules differently in the same build.

I believe you can apply this to `--platforms` today.

**Comment by wpieterse2825** [comment]
Alright, thanks @gregestren will try that thanks for all the help

**Comment by konkers** [comment]
Would this allow one target to depend on the output of another target with a different platform? 

An example of this could be developing a piece of hardware with its own firmware and a companion app.  The firmware and app target different platforms (say an ARM Cortex M0 microcontroller and windows x86_64)  If the app wants to be able to bundle the firmware update, it needs to depend on the output of its compilation.

**Comment by gregestren** [comment]
That should automatically fall out of this, yes.

**Comment by konkers** [comment]
That's super exciting.  Thanks!

**Comment by prestonvanloon** [comment]
Any chance this can bump to P2? Will it make it into bazel v1?

**Comment by gregestren** [comment]
Hi @prestonvanloon - this is sort of a meta-issue. Various other actively prioritized issues together will essentially fulfill this one. 

In my mind prioritizing this specific bug would involve prioritizing the API as suggested in the original description. But depending on your exact needs it should be possible to get a lot of this work already. So I think it's reasonable to expect this for Bazel 1.0 but the details depend on which languages you want support for and your exact API needs.

All that said, what specifically do you want to do?

**Comment by prestonvanloon** [comment]
My mission is to have a target that can generate a tar / zip archive which binaries for multiple platforms. Specifically, with golang with rules_go. 

Currently, we can manually specify each platform tool chain and build the target, but I would like to have some single target that can build the target for all four platforms.
```bash
# //:main is a go_binary 

# What we have today
bazel build --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo //:main
bazel build --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo //:main
bazel build --platforms=@io_bazel_rules_go//go/toolchain:darwin_amd64_cgo //:main
bazel build --platforms=@io_bazel_rules_go//go/toolchain:windows_amd64_cgo //:main

# What we want for bazel v1
bazel build //:distro_zip # genrule tar/zip with the outputs from the target/platforms above.
```

**Comment by gregestren** [comment]
I think this should be straightforward if you write a rule `zipper` using [Starlark transitions](https://docs.bazel.build/versions/master/skylark/config.html#user-defined-transitions) on the `--platforms` flag.

Also see https://github.com/gregestren/snippets/wiki/Configuring-a-cc_binary-differently-than-a-cc_test for similar example code (not quite the same use case but covers the same principle of making a rule build its deps differently)

Starlark transitions even has [support for split deps](https://docs.bazel.build/versions/master/skylark/config.html#defining-12-transitions), which could offer even more concise syntax (you're building the same dep each time). I'm not sure if that's 100% ready but it might be and I encourage exploration. :)

If any of this sounds promising, I suggest a separate issue or bazel-discuss@ thread to back-and-forth on the details (I'm happy to continue offering guidance). If this all looks obviously straightforward to you, please share your results! It'd be awesome to know if this stuff is helping solve real meaningful use cases.

**Comment by gregestren** [comment]
This bug has become too broad, and isn't actionably relevant. Let's follow https://bazel.build/roadmaps/configuration.html for a finer-grained breakdown of priorities, and add more specialized bazel-discuss questions or issues as needed.

**Comment by rajivshah3** [comment]
Does that mean #10061 should be reopened then? It was closed because it would be tracked by this issue (https://github.com/bazelbuild/bazel/issues/10061#issuecomment-551803572)

**Comment by gregestren** [comment]
I see your point. Maybe we should keep this open, then.

My concern is to keep it well-scoped, since "flagless" implies "no flags" and there are many kinds of flags in the most generic sense. We're really only focusing on `--cpu` and `--platforms`.

I'll re-open and adjust the title some.

**Comment by katre** [comment]
Yes, this is close to being doable. I'm working on some high-level proposals now (since [my last attempt](https://github.com/bazelbuild/proposals/blob/main/designs/2022-08-03-platforms-on-targets.md) wasn't workable for a lot of users).

**Comment by sluongng** [comment]
Hey folks, any update on support for multiple target platforms?
Today we are seeing this warning

```
WARNING: --platforms only supports a single target platform: using the first option //platforms:linux_x86_64
```

and the code for this warning points back to this issue.

**Comment by katre** [comment]
Trying to get Bazel to directly support multiple values for `--platforms` is semantically very confusing, so we've made the decision to leave the specific semantics up to rule authors, who know their users' needs and ecosystems better.

To that end, my concrete recommendations are:
1. Rule authors should add `platform` and/or `platforms` attributes to appropriate rules, using the features from https://github.com/bazelbuild/proposals/blob/main/designs/2023-06-08-standard-platform-transitions.md, once we get those released (we're setting up the release process for that now, please imagine the sound of bureaucracy happening in the background).
2. Rule authors can also add custom flags for multi-platform support, like `--android_platforms`, where that makes sense (for example, Android uses this to create a fat APK from a single `android_binary` target).
3. Bazel should fix the `--platforms` flag to be less confusing (#19807):
   1. Make the existing multi-valued flag single-valued, but leave the name.
   2. Actually rename the flag to `--platform` (flag migrations are hard, sorry), leaving `--platforms` as a deprecated legacy name.
   3. Actually remove `--platforms`.

**Comment by Silic0nS0ldier** [comment]
I think I spied a regression in the `--platforms` invalid options handling due to https://github.com/bazelbuild/bazel/commit/81acd2a352387399914ea1ef2a5efa1b4ecc766b#diff-a437b5091032394bcb6af037a2c2ffc063f330a9092144f498edb2feda14d782R252.

The check is performed in `PlatformConfiguration.reportInvalidOptions` which uses `PlatformOptions`, however the instance it receives has `.platforms` already normalised so extraneous platforms are silently ignored (previously a warning was emitted).

**Comment by katre** [comment]
Yes, there's a lot of confusion about the right way to handle this.

**Comment by hisener** [comment]
I think we should keep this feature request open.

**Comment by sluongng** [comment]
@hisener could you please elaborate more on your use case?

This issue is very old and mostly out of date? The community is now in favor of using Bazel Transitions to set the target platform for each individual target.

So let's say if you have a `cc_binary` that you want to build for a specific target, you can do something like this

```
cc_binary(name = "foo", ..., tags = ["manual"])

a_platform_transition_rule(src = ":foo")
```

Popular rules, such as rules_go, offer platform transitions as a direct attribute for more convenience. There are also things like with_cfg to scope down the transitions easier.

So I don't see a reason to keep this one open.

**Comment by hisener** [comment]
When I subscribed to this issue, our use case was using different JDKs for some targets. It's not possible out of the box. We ended up using with_cfg as you mentioned. However, there are a bunch of gotchas, e.g., `env_inherit` doesn't work in some cases.

Thus, I think having something built-in to define a target's intended platform is still valuable.

**Comment by fmeum** [comment]
> However, there are a bunch of gotchas, e.g., env_inherit doesn't work in some cases.

with_cfg should support `env_inherit` and generally aims to shield you from such transition gotchas. Could you file an issue for this over on the with_cfg.bzl repo?

**Comment by hisener** [comment]
> with_cfg should support `env_inherit` and generally aims to shield you from such transition gotchas. Could you file an issue for this over on the with_cfg.bzl repo?

Yes, it's on my list for a while. 😞 I'll try to find some time.

This is what I wrote in October:

> They have a special treatment for env_inherit, but it doesn’t work in our usage because we wrap a function not the java_test native rule (kind:<function _dd_java_test from //rules/java:defs.bzl>). See also https://github.com/fmeum/with_cfg.bzl/blob/178f719f9ffdab3849cdbda259e8debf46c61dda/with_cfg/private/wrapper.bzl#L115-L121.

**Comment by fmeum** [comment]
`env_inherit` on `java_test` is backed by `RunEnvironmentInfo` in Bazel 7 and higher and with_cfg forwards this provider, so I would expect that to work without the hack necessary for native `java_test`.

**Comment by sluongng** [comment]
I think since the direction for configuring targeted platforms has been decided to be using transitions, and viable transition machinery do exist today, it's safe to close this issue to reduce noise (some information from 2018-2019 is no longer applicable to 2026). We can iron out the UX in newer issues that are more focused on the exact problem rather than this broad topic.

There are some remaining TODOs from this issue, though:

> - Actually rename the flag to --platform (flag migrations are hard, sorry), leaving --platforms as a deprecated legacy name.
>
> - Actually remove --platforms.

But other than that, we should move the with_cfg conversation to a different place and close this one.

**Comment by gregestren** [comment]
Agreed on closing for more focused requests. (and thanks for supporting that conclusion @hisener ). 

Acknowledged that there are various ways to kind of achieve this functionality and a canonical best practice API remains elusive.
